### PR TITLE
[posw p1] Implement lock-up for PoSW coinbases

### PR DESCRIPTION
**Pull request by ninjaahhh**
also using `POSW_CONFIG.ENABLED` as the feature flag

### CODE REVIEW

File: quarkchain/cluster/shard_state.py [hunk 1]
```diff
@@ -671,7 +690,8 @@ def add_block(self, block, skip_if_too_old=True):
                     )
                 )
 
-        if self.db.contain_minor_block_by_hash(block.header.get_hash()):
+        header_hash = block.header.get_hash()
```
Comment (qizhou) [inline quarkchain/cluster/shard_state.py]: I don't prefer pre-mature optimization, especially such variable is used multiple places in the method and it is hard to find its context when reading its references.

File: quarkchain/cluster/shard_state.py [hunk 2]
```diff
@@ -119,36 +120,48 @@ def init_from_root_block(self, root_block):
...
         self.__rewrite_block_index_to(
-            self.db.get_minor_block_by_hash(self.header_tip.get_hash()),
-            add_tx_back_to_queue=False,
+            self.db.get_minor_block_by_hash(header_tip_hash), add_tx_back_to_queue=False
         )
 
-    def __create_evm_state(self):
-        return EvmState(
+    def __create_evm_state(
+        self, root_hash: Optional[bytes], header_hash: Optional[bytes]
```
Comment (qizhou) [inline quarkchain/cluster/shard_state.py]: I would suggest renaming root_hash to trie_root_hash to avoid confusion with root block hash

Comment (qizhou) [reply quarkchain/cluster/shard_state.py]: or evm_state_root_hash

File: quarkchain/evm/messages.py [hunk 3]
```diff
@@ -366,6 +366,7 @@ def __init__(self, state, tx):
         self.reset_storage = state.reset_storage
         self.tx_origin = tx.sender if tx else b"\x00" * 20
         self.tx_gasprice = tx.gasprice if tx else 0
+        self.posw_disallow_list = state.posw_disallow_list
```
Comment (qcgg) [inline quarkchain/evm/messages.py]: nit:
following the same pattern here, we can add a `is_locked` function to state and then pass that function to VMExt.

Comment (ninjaahhh) [reply quarkchain/evm/messages.py]: I think it's more of a functional approach, and mostly a coding style thing. personally I think passing the list to VM is rather more straightforward and easier to read (unless in the future we would have a more complicated logic on disallowing senders)

**Comment by ninjaahhh** [comment]
part of #306

**Pull request merged**